composition-js/src/merging/merge.ts

```diff
@@ -3606,4 +3689,344 @@ class Merger {
+        for (const targetTypeName of targetTypeNames) {
+          // we need to verify against all possible contexts
+          const targetType = this.schema.type(targetTypeName) as CompositeType;
+          assert(targetType, 'Context references valid type in the schema');
+          try {
+            const requirementsOnContextType = this.authRequirementsOnElement(targetType);
+            if (!authRequirementOnContext.satisfies(requirementsOnContextType)) {
+              const msg = `Field "${field.coordinate}" does not specify necessary @authenticated, @requiresScopes `
+                  + `and/or @policy auth requirements to access the transitive data in context ${name} from @fromContext selection set.`;
+              throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+            }
+
+            const contextSelectionSet = parseSelectionSet({parentType: targetType, source: contextSelection});
+            this.verifyAuthRequirementsOnSelectionSet(authRequirementOnContext, contextSelectionSet);
+          } catch (e) {
+            if (!(e instanceof GraphQLError)) {
+              throw e;
+            }
+            // target subgraph info should always be provided but just in case
+            const enumSubgraphValue = joinDirectiveOnFromContext.arguments().graph;
+            const subgraph = enumSubgraphValue ? this.joinSpecNamesToSubgraphNames.get(enumSubgraphValue) : undefined;
+            if (subgraph) {
+              errors.push(addSubgraphToError(e, subgraph));
+            } else {
+              errors.push(e);
+            }
+          }
+        }
+      }
+    }
+    return errors;
+  }
+
+  private authRequirementsOnElement(element: NamedType | FieldDefinition<ObjectType>): AuthRequirementsOnElement | undefined {
+    const requirements = new AuthRequirementsOnElement();
+    if (this.authenticatedDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.authenticatedDirective)?.[0];
+      if (appliedDirective) {
+        requirements.isAuthenticated = true;
+      }
+    }
+
+    if (this.requiresScopesDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.requiresScopesDirective)?.[0];
+      if (appliedDirective) {
+        const { scopes } = appliedDirective.arguments();
+        requirements.scopes = scopes;
+      }
+    }
+
+    if (this.policyDirective) {
+      const appliedDirective = element.appliedDirectivesOf(this.policyDirective)?.[0];
+      if (appliedDirective) {
+        const { policies } = appliedDirective.arguments();
+        requirements.policies = policies;
+      }
+    }
+
+    if (requirements.isAuthenticated || requirements.scopes || requirements.policies) {
+      return requirements;
+    } else {
+      return;
+    }
+  }
+
+  private verifyAuthRequirementsOnSelectionSet(authRequirements: AuthRequirements, selectionSet: SelectionSet) {
+    const parentType = this.schema.type(selectionSet.parentType.name);
```

Comment:
Since the selection set started out being parsed on the merged schema, I'm not sure why we're refetching the parent type from the merged schema here.

Reply:
Indeed this is a leftover from some previous versions.